examples/openai_tools_calls.py

```diff
@@ -0,0 +1,241 @@
+                "temperature": temperature,
+                "unit": unit
+            })
+        elif "paris" in location.lower():
+            temperature = "72" if unit.lower() == "fahrenheit" else "22"
+            return json.dumps({
+                "location": "Paris",
+                "temperature": temperature,
+                "unit": unit
+            })
+    return json.dumps({"location": str(location), "temperature": "unknown"})
+
+
+def get_current_date_utc():
+    print("Calling get_current_date_utc client side.")
+    return datetime.datetime.now(datetime.timezone.utc).strftime(
+        "The current UTC datetime is (day: %A, date (day/month/year): %d/%m/%Y, time: %H:%M)."
+    )
+
+
+def run_conversation(question: str, tool_choice_param):
+    # Step 1: send the conversation and available functions to the model
+    # messages = [{"role": "user", "content": "What's the weather like in San Francisco, Tokyo, and Paris?"}]
+    messages = [{"role": "user", "content": question}]
+    tools = [{
+        "type": "function",
+        "function": {
+            "name": "get_current_weather",
+            "description": "Get the current weather in a given location",
+            "parameters": {
+                "type": "object",
+                "properties": {
+                    "location": {
+                        "type":
+                        "string",
+                        "description":
+                        "The city and state, e.g. San Francisco, CA as a string",
+                    },
+                    "unit": {
+                        "type": "string",
+                        "enum": ["celsius", "fahrenheit"]
+                    },
+                },
+                "required": ["location"],
+            },
+        },
+    }, {
+        "type": "function",
+        "function": {
+            "name": "get_current_date_utc",
+            "description": "Get the current UTC time",
+        },
+    }]
+    response = client.chat.completions.create(model=model,
+                                              messages=messages,
+                                              tools=tools,
+                                              stream=stream,
+                                              tool_choice=tool_choice_param,
+                                              temperature=temperature,
+                                              extra_body=EXTRA_BODY_OPENAI)
+    response_message = ""
+    tool_calls = []
+    if stream:
+        text_message = ""
+        for chunk in response:
+            if chunk.choices[0].finish_reason is not None:
+                if chunk.choices[0].finish_reason == "tool_calls":
+                    tool_calls += chunk.choices[0].delta.tool_calls
+                    # print("TEST : %s" % chunk.choices[0].delta.tool_calls)
+                break
+            if chunk.choices[0].delta.content is not None:
+                text_message += chunk.choices[0].delta.content
```

Comment:
For what it's worth, probably the better way to handle this is to handle the response stream one chunk or token at a time. If you get a token indicating a tool call (such as `<tool_call>`) at the start of the response then you want to wait for the entire response from the LLM so that you can invoke the tool. if you get a non-meta or non-control token (e.g. a normal streaming text chat response) then you probably want to start showing the streaming tokens to the user immediately, avoiding the latency of the entire response. But, This is also an _example_ so I'm aware it's not necessary for it to be optimized.